alert/expiring.py

```diff
@@ -0,0 +1,86 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# Search for regression in a histogram dump directory produced by the
+# node exporter.
+
+import json
+import urllib2
+from datetime import datetime, timedelta
+
+from bs4 import BeautifulSoup
+from mail import send_ses
+from version_compare import version_compare
+
+HISTOGRAMS_FILE = "Histograms.json"
+EMAIL_TIME_BEFORE = timedelta(weeks=1)
+FROM_ADDR = "telemetry-alert@mozilla.com"
+
+def get_future_release_dates():
+    # scrape for future release date tables
+    response = json.loads(urllib2.urlopen("https://wiki.mozilla.org/api.php?action=parse&format=json&page=RapidRelease/Calendar").read())
+    soup = BeautifulSoup(response["parse"]["text"]["*"])
+    table = soup.find(id="Future_branch_dates").find_parent("h2").find_next_sibling("table")
+    result = {}
+    for i, row in enumerate(table.find_all("tr")):
+        if i == 0: continue # skip the header row
+        version = list(row.find_all("td"))[-1].string.strip().replace("Firefox ", "")
+        version = version.split(".")[0] # only keep the major version
+        date_string = list(row.find_all("th"))[-1].string.strip(" \t\r\n*")
+        try: result[version] = datetime.strptime(date_string, "%Y-%m-%d")
+        except ValueError: pass
+    return result
+
+def is_expiring(histogram_entry, now, release_dates):
+    # check if the expiration version is unknown, invalid, or far into the future
+    expiry = histogram_entry.get("expires_in_version", "never").strip()
+    if expiry in {"never", "default"}: return False
+    try:
+        max_release = next(release_dates.iterkeys())
+        for release in release_dates.iterkeys():
+            if version_compare(max_release, release) < 0:
+                max_release = release
+        if version_compare(expiry, max_release) > 0: return False
```

Comment:
This is still incorrect. You are not accounting correctly for the various trees when parsing future branch dates. A histogram might expire for example in nightly (e.g. 40.0a1), in beta (41.0b1), in aurora (e.g. 40.0a2), or in release (41.0), depending on the version number.